UnitTests/BraintreeThreeDSecureTests/BTThreeDSecureClient_Tests.swift

```diff
@@ -253,20 +258,37 @@ class BTThreeDSecureClient_Tests: XCTestCase {
         waitForExpectations(timeout: 2)
         XCTAssertTrue(mockAPIClient.postedAnalyticsEvents.contains(BTThreeDSecureAnalytics.verifyFailed))
     }
+
+    func testStartPaymentFlow_whenAmountIsNil_throwsError() {
+        mockAPIClient.cannedConfigurationResponseBody = BTJSON(value: [
+            "threeDSecure": ["cardinalAuthenticationJWT": "FAKE_JWT"],
+            "assetsUrl": "http://assets.example.com"
+        ] as [String: Any])
+
+        let request = BTThreeDSecureRequest()
+        request.amount = nil
+
+        let expectation = expectation(description: "Callback envoked")
+
+        client.startPaymentFlow(request) { result, error in
+            XCTAssertNil(result)
+            XCTAssertEqual(error?.localizedDescription, "BTThreeDSecureRequest amount can not be nil or NaN.")
+            expectation.fulfill()
+        }
+
+        waitForExpectations(timeout: 2)
+        XCTAssertTrue(mockAPIClient.postedAnalyticsEvents.contains(BTThreeDSecureAnalytics.verifyFailed))
+    }
     
-    func testStartPayment_returnsError_whenAmountIsMissing() {
+    func testStartPayment_whenNoBodyReturned_returnsAnError() {
         threeDSecureRequest = BTThreeDSecureRequest()
         threeDSecureRequest.nonce = "fake-card-nonce"
         threeDSecureRequest.threeDSecureRequestDelegate = mockThreeDSecureRequestDelegate
 
         let expectation = expectation(description: "willCallCompletion")
+        mockAPIClient.cannedConfigurationResponseBody = mockConfiguration
 
-        mockAPIClient.cannedConfigurationResponseBody = BTJSON(value: [
-            "threeDSecure": ["cardinalAuthenticationJWT": "FAKE_JWT"],
-            "assetsUrl": "http://assets.example.com"
-        ] as [String: Any])
-        let client = BTThreeDSecureClient(apiClient: mockAPIClient)
-
+        let client = MockThreeDSecureClient(apiClient: mockAPIClient)
```

Comment:
@jaxdesmarais A general rule of thumb - you don't want to mock the system/subject under test that you're testing. So in this file, the SUT is `BTThreeDSecureClient`. If we do this, then we aren't actually testing `BTThreeDSecureClient`.

I think the cleanest approach here is to mock only the dependencies of `BTThreeDSecureClient` that need to be: for example to avoid network calls, DBs, or other external dependencies.

Check out[ this branch](https://github.com/braintree/braintree_ios/compare/fix-flakey-tests...mock-cardinal-session) which mocks `CardinalSession` directly via a protocol.

Reply:
That makes sense. Yeah I wasn't sure where the line was with mocking the SUT in this case. But mocking `CardinalSession` via a protocol makes sense to me. I'll merge in that branch.

Agree we can punt on adding tests for `BTThreeDSecureV2Provider` to get us unblocked for now on test failures.

Reply:
Merged in here: https://github.com/braintree/braintree_ios/pull/1067/commits/faf81dd64913c08f6ae29371d4cb6e13c11fa02e

Reply:
Sweet! Yeh I went with `CardinalSessionTestable` for the protocol name. We should pick a naming convention for these kind of protocols, stick to it, and add it to our style guide. We have a few in PPCP too. I think some are just `<THING>Protocol`. I like `Testable` 🤷

Reply:
Would `Testable` only be for protocols used for testing/mocking? I like that naming for those purposes but feels a little odd for things that aren't used for mocking/testing. If that makes sense. Agree we should have a standard though just not sure if we need 1 for testing/mocking and 1 for everything else/merchant facing things/etc